File: backend/industrial_standards.py
# Industrial Standards Configuration
# Master source for all industrial thresholds and standards
# Updated to follow ISO 10816-3 and NEMA standards

# Temperature thresholds (Celsius) - NEMA Class B Insulation Standards
TEMPERATURE_THRESHOLDS = {
    "good": {"min": 20, "max": 70},           # Normal operating range
    "satisfactory": {"min": 70, "max": 85},   # Acceptable but monitor
    "unsatisfactory": {"min": 85, "max": 105}, # Requires attention
    "unacceptable": {"min": 105, "max": 200}   # Immediate action required
}

# Vibration thresholds (mm/s RMS) - ISO 10816-3 Standard
VIBRATION_THRESHOLDS = {
    "good": {"min": 0, "max": 1.8},          # Zone A: Excellent condition
    "satisfactory": {"min": 1.8, "max": 4.5}, # Zone B: Acceptable operation
    "unsatisfactory": {"min": 4.5, "max": 11.2}, # Zone C: Restricted operation
    "unacceptable": {"min": 11.2, "max": 50.0}   # Zone D: Unacceptable operation
}

# Speed thresholds (RPM) - Industrial motor operation standards
SPEED_THRESHOLDS = {
    "good": {"min": 1150, "max": 1200},       # Normal operating range
    "satisfactory": {"min": 1200, "max": 1300}, # Slightly elevated but acceptable
    "unsatisfactory": {"min": 1300, "max": 1450}, # High speed - monitor closely
    "unacceptable": {"min": 1450, "max": 5000}    # Dangerous overspeed
}

# Combined thresholds for easy access
INDUSTRIAL_STANDARDS = {
    "temperature": TEMPERATURE_THRESHOLDS,
    "vibration": VIBRATION_THRESHOLDS,
    "speed": SPEED_THRESHOLDS
}

# Industry standard condition levels
CONDITION_LEVELS = ["good", "satisfactory", "unsatisfactory", "unacceptable"]

def get_threshold(parameter, condition):
    """Get threshold range for a specific parameter and condition"""
    return INDUSTRIAL_STANDARDS.get(parameter, {}).get(condition, {})

def get_all_thresholds():
    """Get all industrial standards thresholds"""
    return INDUSTRIAL_STANDARDS

def classify_value(parameter, value):
    """Classify a sensor value based on industry standards (ISO 10816-3 & NEMA)"""
    thresholds = INDUSTRIAL_STANDARDS.get(parameter, {})
    
    if not thresholds:
        return "unknown"
    
    # Check each condition level in order
    for level in CONDITION_LEVELS:
        if level in thresholds:
            range_info = thresholds[level]
            if range_info["min"] <= value < range_info["max"]:
                return level
    
    # If value exceeds all ranges, return most severe
    return "unacceptable"

def get_condition_score(temperature, vibration, speed):
    """Calculate overall condition score based on all parameters using industry standards"""
    scores = {
        "good": 0,
        "satisfactory": 1,
        "unsatisfactory": 2,
        "unacceptable": 3
    }
    
    temp_condition = classify_value("temperature", temperature)
    vib_condition = classify_value("vibration", vibration)
    speed_condition = classify_value("speed", speed)
    
    # Get the worst condition (highest score)
    conditions = [temp_condition, vib_condition, speed_condition]
    condition_scores = [scores.get(cond, 3) for cond in conditions]
    max_score = max(condition_scores)
    
    # Return the condition level corresponding to the worst score
    for condition, score in scores.items():
        if score == max_score:
            return condition
    
    return "unacceptable"

def get_detailed_analysis(temperature, vibration, speed):
    """Get detailed analysis of all parameters with industry standard classifications"""
    temp_condition = classify_value("temperature", temperature)
    vib_condition = classify_value("vibration", vibration)
    speed_condition = classify_value("speed", speed)
    overall_condition = get_condition_score(temperature, vibration, speed)
    
    return {
        "overall_condition": overall_condition,
        "temperature": {
            "value": temperature,
            "condition": temp_condition,
            "unit": "°C",
            "standard": "NEMA Class B"
        },
        "vibration": {
            "value": vibration,
            "condition": vib_condition,
            "unit": "mm/s RMS",
            "standard": "ISO 10816-3"
        },
        "speed": {
            "value": speed,
            "condition": speed_condition,
            "unit": "RPM",
            "standard": "Industrial Practice"
        }
    }

# Industry standards reference information
STANDARDS_INFO = {
    "temperature": {
        "standard": "NEMA MG-1 / IEC 60034-1",
        "description": "Class B insulation motor temperature limits",
        "reference": "Maximum hot spot temperature 130°C for Class B insulation"
    },
    "vibration": {
        "standard": "ISO 10816-3:2009",
        "description": "Industrial machinery vibration evaluation",
        "reference": "For machines >15kW, 120-15000 RPM, Group 2 classification"
    },
    "speed": {
        "standard": "Industry Best Practice",
        "description": "Typical 4-pole, 60Hz industrial motor operation",
        "reference": "Synchronous speed 1800 RPM, typical load operation 1150-1200 RPM"
    }
}

if __name__ == "__main__":
    # Test the industrial standards functions
    print("🏭 Testing Industrial Standards Module")
    print("=" * 50)
    
    # Display standards information
    print("\n📋 Industry Standards Reference:")
    for param, info in STANDARDS_INFO.items():
        print(f"\n{param.upper()}:")
        print(f"  Standard: {info['standard']}")
        print(f"  Description: {info['description']}")
        print(f"  Reference: {info['reference']}")
    
    print("\n" + "=" * 50)
    print("🧪 Test Classifications:")
    
    # Test cases
    test_cases = [
        (65.0, 1.5, 1180.0, "Normal operation"),
        (78.0, 3.2, 1250.0, "Satisfactory condition"),
        (92.0, 6.8, 1380.0, "Unsatisfactory condition"),
        (115.0, 15.0, 1520.0, "Unacceptable condition")
    ]
    
    for temp, vib, speed, description in test_cases:
        analysis = get_detailed_analysis(temp, vib, speed)
        print(f"\n{description}:")
        print(f"  Values: T={temp}°C, V={vib}mm/s, S={speed}RPM")
        print(f"  Overall: {analysis['overall_condition'].upper()}")
        print(f"  Temperature: {analysis['temperature']['condition']} ({analysis['temperature']['standard']})")
        print(f"  Vibration: {analysis['vibration']['condition']} ({analysis['vibration']['standard']})")
        print(f"  Speed: {analysis['speed']['condition']} ({analysis['speed']['standard']})")
    
    print(f"\n{'=' * 50}")
    print("✅ Industrial standards module working correctly!")

File: frontend/src/constants/industrialStandards.js
// Industrial Standards Constants
// Static constants for industrial thresholds - these should match backend/industrial_standards.py

// Temperature thresholds (Celsius)
export const TEMPERATURE_THRESHOLDS = {
  NORMAL: { MIN: 20, MAX: 65 },
  WARNING: { MIN: 65, MAX: 85 },
  CRITICAL: { MIN: 85, MAX: 120 }
};

// Vibration thresholds (mm/s RMS)
export const VIBRATION_THRESHOLDS = {
  NORMAL: { MIN: 0, MAX: 3.0 },
  WARNING: { MIN: 3.0, MAX: 7.0 },
  CRITICAL: { MIN: 7.0, MAX: 15.0 }
};

// Speed thresholds (RPM)
export const SPEED_THRESHOLDS = {
  NORMAL: { MIN: 800, MAX: 1150 },
  WARNING: { MIN: 1150, MAX: 1350 },
  CRITICAL: { MIN: 1350, MAX: 2000 }
};

// Combined thresholds object
export const INDUSTRIAL_THRESHOLDS = {
  temperature: {
    normal: TEMPERATURE_THRESHOLDS.NORMAL,
    warning: TEMPERATURE_THRESHOLDS.WARNING,
    critical: TEMPERATURE_THRESHOLDS.CRITICAL
  },
  vibration: {
    normal: VIBRATION_THRESHOLDS.NORMAL,
    warning: VIBRATION_THRESHOLDS.WARNING,
    critical: VIBRATION_THRESHOLDS.CRITICAL
  },
  speed: {
    normal: SPEED_THRESHOLDS.NORMAL,
    warning: SPEED_THRESHOLDS.WARNING,
    critical: SPEED_THRESHOLDS.CRITICAL
  }
};

// Condition severity levels
export const CONDITION_LEVELS = {
  NORMAL: 'normal',
  WARNING: 'warning',
  CRITICAL: 'critical',
  UNKNOWN: 'unknown'
};

// Condition colors for UI
export const CONDITION_COLORS = {
  [CONDITION_LEVELS.NORMAL]: '#10B981',    // Green
  [CONDITION_LEVELS.WARNING]: '#F59E0B',   // Amber
  [CONDITION_LEVELS.CRITICAL]: '#EF4444',  // Red
  [CONDITION_LEVELS.UNKNOWN]: '#6B7280'    // Gray
};

// Condition icons
export const CONDITION_ICONS = {
  [CONDITION_LEVELS.NORMAL]: '✅',
  [CONDITION_LEVELS.WARNING]: '⚠️',
  [CONDITION_LEVELS.CRITICAL]: '🚨',
  [CONDITION_LEVELS.UNKNOWN]: '❓'
};

// Parameter units
export const PARAMETER_UNITS = {
  temperature: '°C',
  vibration: 'mm/s',
  speed: 'RPM'
};

// Parameter display names
export const PARAMETER_NAMES = {
  temperature: 'Temperature',
  vibration: 'Vibration',
  speed: 'Speed'
};

// Utility functions
export const classifyValue = (parameter, value) => {
  const thresholds = INDUSTRIAL_THRESHOLDS[parameter];
  if (!thresholds) return CONDITION_LEVELS.UNKNOWN;

  if (value <= thresholds.normal.max) {
    return CONDITION_LEVELS.NORMAL;
  } else if (value <= thresholds.warning.max) {
    return CONDITION_LEVELS.WARNING;
  } else {
    return CONDITION_LEVELS.CRITICAL;
  }
};

export const getConditionColor = (condition) => {
  return CONDITION_COLORS[condition] || CONDITION_COLORS[CONDITION_LEVELS.UNKNOWN];
};

export const getConditionIcon = (condition) => {
  return CONDITION_ICONS[condition] || CONDITION_ICONS[CONDITION_LEVELS.UNKNOWN];
};

export const calculateOverallCondition = (temperature, vibration, speed) => {
  const scores = {
    [CONDITION_LEVELS.NORMAL]: 0,
    [CONDITION_LEVELS.WARNING]: 1,
    [CONDITION_LEVELS.CRITICAL]: 2
  };
  
  const tempCondition = classifyValue('temperature', temperature);
  const vibCondition = classifyValue('vibration', vibration);
  const speedCondition = classifyValue('speed', speed);
  
  const totalScore = scores[tempCondition] + scores[vibCondition] + scores[speedCondition];
  
  if (totalScore >= 4) {
    return CONDITION_LEVELS.CRITICAL;
  } else if (totalScore >= 2) {
    return CONDITION_LEVELS.WARNING;
  } else {
    return CONDITION_LEVELS.NORMAL;
  }
};

export const getThresholdDisplayText = (parameter, condition) => {
  const threshold = INDUSTRIAL_THRESHOLDS[parameter]?.[condition];
  if (!threshold) return 'N/A';

  const unit = PARAMETER_UNITS[parameter] || '';
  
  if (condition === CONDITION_LEVELS.NORMAL) {
    return `≤ ${threshold.max}${unit}`;
  } else if (condition === CONDITION_LEVELS.CRITICAL) {
    return `> ${threshold.min}${unit}`;
  } else {
    return `${threshold.min} - ${threshold.max}${unit}`;
  }
};

// Export default object with all constants
export default {
  TEMPERATURE_THRESHOLDS,
  VIBRATION_THRESHOLDS,
  SPEED_THRESHOLDS,
  INDUSTRIAL_THRESHOLDS,
  CONDITION_LEVELS,
  CONDITION_COLORS,
  CONDITION_ICONS,
  PARAMETER_UNITS,
  PARAMETER_NAMES,
  classifyValue,
  getConditionColor,
  getConditionIcon,
  calculateOverallCondition,
  getThresholdDisplayText
};